### PR TITLE
new features and overall cleanup

### PR DESCRIPTION
**Pull request by jimblue**
Hi @naucon !

Since your last update about language, I was curious to see how you did it.

And by looking at your code I saw many little things to improve.
So I decide to work on it this morning:

- Add option to enabled/disable CSS loading (Use built in CSS).
- Add option to enabled/disable JS loading (Use built in JS).
- Add option to choose inline js position (Load js options in bottom).
- Add option to choose a custom event to delay Cookie consent initialisation.
- Improved loading event of js settings (more reliable than listening on load event).
- Improved admin frontend form to choose options (add color picker and section).
- Cleanup PHP code (before all the Grav config was sent to the template)
- Cleanup blueprint and template from unnecessary message (they can be customised with languages.yaml since v0.2.0)

I've test it and it's working like a charm.

Looking forward the merge 😄 !

Cheers

### CODE REVIEW

**Comment by jimblue** [comment]
Hi @naucon did you see the PR? 😄

File: blueprints.yaml [hunk 1]
```diff
@@ -23,66 +23,81 @@ form:
         0: Disabled
       validate:
         type: bool
```
Comment (naucon) [inline blueprints.yaml:25]: most of the users don't need translations and would like to customize the cookie consent texts over the admin interface. Also it would breaks compatibility to older versions. Therefore I would like to keep `content_message`, `content_dismiss`, `content_link` and `content_href`.

File: cookieconsent.php [hunk 2]
```diff
@@ -58,12 +58,30 @@ public function onTwigTemplatePaths()
...
-        $config = $this->config->toArray();
+        // Configuration
+        $builtInCss = $cookieConsentOptions['built_in_css'];
+        $builtInJs = $cookieConsentOptions['built_in_js'];
+        $loadJsOptionsBottom = $cookieConsentOptions['load_js_options_bottom'];
 
-        $this->grav['assets']->addCss("//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.0.3/cookieconsent.min.css");
-        $this->grav['assets']->addJs("//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.0.3/cookieconsent.min.js");
+        // Load cookie consent CSS if enabled
+        if (!$this->isAdmin() && $builtInCss) {
```
Comment (naucon) [inline cookieconsent.php:70]: the `isAdmin()` in the if-condition is not required because it is already excluded the `onPluginsInitialized` method
```
    /**
     * Initialize the plugin
     */
    public function onPluginsInitialized()
    {
        // Don't proceed if we are in the admin plugin
        if ($this->isAdmin()) {
            return;
        }
        // Enable the main event we are interested in
        $this->enable([
            'onTwigTemplatePaths' => ['onTwigTemplatePaths', 0],
            'onTwigSiteVariables' => ['onTwigSiteVariables', 0],
        ]);
    }
```

File: cookieconsent.php [hunk 3]
```diff
@@ -58,12 +58,30 @@ public function onTwigTemplatePaths()
...
 
-        $this->grav['assets']->addInlineJs($twig->twig->render('partials/cookieconsent.html.twig', array('config' => $config)));
+          // Add cookie consent CSS style
+          $assets->addCss('//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.0.3/cookieconsent.min.css');
+        }
+
+        // Load cookie consent JS if enabled
+        if (!$this->isAdmin() && $builtInJs) {
```
Comment (naucon) [inline cookieconsent.php:77]: Looks like this if-condition is identical to the one in line 70.

File: blueprints.yaml [hunk 4]
```diff
@@ -23,66 +23,81 @@ form:
...
+      highlight: 1
+      default: 1
+      options:
+        1: Enabled
+        0: Disabled
       validate:
-        type: text
-    content_href:
+        type: bool
+    custom_document_event:
```
Comment (naucon) [inline blueprints.yaml:60]: I am interested why did you need this custom event? Did you had any trouble with the theme integration?

File: blueprints.yaml [hunk 5]
```diff
@@ -23,66 +23,81 @@ form:
...
-    content_message:
-      type: text
-      label: Message
-      help: Message on the banner
-      placeholder: "This website uses cookies to ensure you get the best experience on our website."
+    assets:
+      type: section
+      title: Assets
+      underline: true
+    built_in_css:
```
Comment (naucon) [inline blueprints.yaml:30]: I am wunder whats that option is all about.
Look like when is option is turned off no JS and CSS is added and no template is rendered. basically its looks like disabling the hole plugin but with a extra option.

```
        // Load cookie consent JS if enabled
        if (!$this->isAdmin() && $builtInJs) {
          // Add cookie consent JS main script
          $assets->addJs('//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.0.3/cookieconsent.min.js');
          // Add cookie consent JS settings
          $jsSettings = $twig->twig->render('partials/cookieconsent.html.twig', $cookieConsentOptions);
          $group = $loadJsOptionsBottom ? 'bottom' : null;
          $assets->addInlineJs($jsSettings, null, $group);
        }
```

**Review by naucon — COMMENTED** [review]
Hi jimblue,

first thank you for your pull request and your work to improve this plugin.

I added some comments on the diff where I see some problems. I would be grateful if you could give me your point of view on this things to find a mutual solution.

But breaking compatibility is out of the question.

File: blueprints.yaml (hunk 5 above)
Comment (jimblue) [reply blueprints.yaml:30]: Because some people use bundle bundler like webpack to optimise loading performance.
This is what I do, I import on my own the js and css with NPM and bundle them with the rest of my assets.

The only thing that's needed in that case is the JS settings.

File: blueprints.yaml (hunk 4 above)
Comment (jimblue) [reply blueprints.yaml:60]: Because cookie consent will popup as soon as it's loaded it can break some design.
For exemple my website have a loading page with animations while page is loading.

Without custom event cookie consent just popup on top of this page and that's ugly...

File: cookieconsent.php (hunk 2 above)
Comment (jimblue) [reply cookieconsent.php:70]: That's true! We can't remove this condition.

File: cookieconsent.php (hunk 3 above)
Comment (jimblue) [reply cookieconsent.php:77]: idem

File: blueprints.yaml (hunk 1 above)
Comment (jimblue) [reply blueprints.yaml:25]: I remove them because it can be done with translation files. But I understand your point. Of course you can put them back

**Comment by jimblue** [comment]
Hi @naucon, I've reviewed all your comment, what do you think?

File: blueprints.yaml (hunk 5 above)
Comment (naucon) [reply blueprints.yaml:30]: My goal was a plug and play solution. Install the plugin. Maybe change colors, position and text over the admin interface and your good to go. No coding required.

If you have your own theme, using webpack to bundle cookieconsent assets, loading the cookieconsent over a customer event - there is not much left for the plugin to do. And if you disable `buildInCss` and `buildInJs` the plugin actually do nothing at all. I don't see the point in installing the plugin in this case.

**Comment by naucon** [comment]
Hi @jimblue, thank you for your answers. Sorry for letting you wait. Only on the weekend I have the time to take care about my open source projects.

At the moment I would like to add the
- color picker
- custom event
- loading option (maybe rewriting it a bit)

For disabling the `buildInJs` (JS file + Template) or `buildInCss` (CSS file) I don't the the point. But I am open for convincing arguments.

**Comment by jimblue** [comment]
No worries for the delay.

About disabling the buildInJs (JS file + Template) and/or buildInCss (CSS file) it's important for people using bundler like webpack. (I will add JS and CSS to my package myself... if I can't disable them in your plugin they will be loaded twice...)

**Comment by jimblue** [comment]
Hi @naucon, any update about merging hit PR?
